tests/server.py

```diff
@@ -602,30 +602,48 @@ def _make_test_homeserver_synchronous(server: HomeServer) -> None:
     for database in server.get_datastores().databases:
         pool = database._db_pool
 
-        def runWithConnection(func, *args, **kwargs):
-            return threads.deferToThreadPool(
+        async def runWithConnection(
+            func: Callable[..., R],
+            *args: Any,
+            db_autocommit: bool = False,
+            isolation_level: Optional[int] = None,
+            **kwargs: Any,
+        ) -> R:
```

Comment:
General comment: I think I've toyed with using a ParamSpec here before (or maybe the equivalent function in Synapse). But this gets tricky because that only works if `*args` and `**kwargs` are literally next to each other the function signature.

(Not proposing we change this here.)